File: leetcode/math/1924_Erect_the_Fence_II/codes/welzl.cpp
#include <bits/stdc++.h>
using namespace std;

struct Point { double x, y; };

double dist(const Point &a, const Point &b) {
    return hypot(a.x - b.x, a.y - b.y);
}
Point mid(const Point &a, const Point &b) {
    return {(a.x + b.x) / 2.0, (a.y + b.y) / 2.0};
}
pair<Point,double> circleFrom(vector<Point> R) {
    if (R.empty()) return {{0,0}, 0};
    if (R.size() == 1) return {R[0], 0};
    if (R.size() == 2) {
        Point c = mid(R[0], R[1]);
        return {c, dist(R[0], R[1]) / 2.0};
    }
    Point A = R[0], B = R[1], C = R[2];
    double a = B.x - A.x, b = B.y - A.y;
    double c = C.x - A.x, d = C.y - A.y;
    double e = a * (A.x + B.x) + b * (A.y + B.y);
    double f = c * (A.x + C.x) + d * (A.y + C.y);
    double g = 2 * (a * (C.y - B.y) - b * (C.x - B.x));
    if (fabs(g) < 1e-12) return {{0,0}, numeric_limits<double>::infinity()};
    double cx = (d * e - b * f) / g;
    double cy = (a * f - c * e) / g;
    return {{cx, cy}, dist({cx, cy}, A)};
}
bool inCircle(Point p, Point c, double r) {
    return dist(p, c) <= r + 1e-12;
}
pair<Point,double> welzl(vector<Point> &P, vector<Point> R, int n) {
    if (n == 0 || R.size() == 3) return circleFrom(R);
    int idx = rand() % n;
    swap(P[idx], P[n - 1]);
    Point p = P[n - 1];
    auto circle = welzl(P, R, n - 1);
    if (inCircle(p, circle.first, circle.second)) return circle;
    R.push_back(p);
    return welzl(P, R, n - 1);
}
class Solution {
public:
    vector<double> outerTrees(vector<vector<int>>& trees) {
        set<pair<int,int>> s;
        vector<Point> pts;
        for (auto &t : trees) {
            if (s.insert({t[0], t[1]}).second) pts.push_back({(double)t[0], (double)t[1]});
        }
        random_shuffle(pts.begin(), pts.end());
        auto res = welzl(pts, {}, pts.size());
        return {res.first.x, res.first.y, res.second};
    }
};

File: leetcode/math/1924_Erect_the_Fence_II/codes/welzl.py
from typing import List
import random
import math

class Solution:
    def outerTrees(self, trees: List[List[int]]) -> List[float]:
        def dist(p1, p2):
            return math.hypot(p1[0] - p2[0], p1[1] - p2[1])

        def mid(p1, p2):
            return ((p1[0] + p2[0]) / 2, (p1[1] + p2[1]) / 2)

        def circle_from(points):
            if not points:
                return (0.0, 0.0), 0.0

            if len(points) == 1:
                return points[0], 0.0

            if len(points) == 2:
                c = mid(points[0], points[1])
                r = dist(points[0], points[1]) / 2
                return c, r

            x1, y1 = points[0]
            x2, y2 = points[1]
            x3, y3 = points[2]

            A = x2 - x1
            B = y2 - y1
            C = x3 - x1
            D = y3 - y1
            E = A * (x1  +  x2)  +  B * (y1  +  y2)
            F = C * (x1  +  x3)  +  D * (y1  +  y3)
            G = 2 * (A * (y3 - y2) - B * (x3 - x2))

            if abs(G) < 1e-12:
                return (0.0, 0.0), float('inf')

            cx = (D * E - B * F) / G
            cy = (A * F - C * E) / G
            r = dist((cx, cy), points[0])
            return (cx, cy), r

        def is_in_circle(p, c, r):
            return dist(p, c) <= r  +  1e-12

        def welzl(P, R, n):
            if n == 0 or len(R) == 3:
                return circle_from(R)
            idx = random.randrange(n)
            P[idx], P[n - 1] = P[n - 1], P[idx]
            p = P[n - 1]
            c, r = welzl(P, R, n - 1)
            if is_in_circle(p, c, r):
                return c, r
            return welzl(P, R + [p], n - 1)

        points = list(set(map(tuple, trees)))
        random.shuffle(points)
        center, radius = welzl(points, [], len(points))
        return [center[0], center[1], radius]
